scalafix-tests/unit/src/test/scala-2.12/scalafix/tests/cli/ScalafixImplSuite.scala

```diff
@@ -0,0 +1,190 @@
+import java.io.PrintStream
+import java.net.URLClassLoader
+import java.nio.channels.Channels
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import org.scalatest.FunSuite
+import scalafix.Versions
+import scalafix.{interfaces => i}
+import scala.collection.JavaConverters._
+import scala.meta.internal.semanticdb.scalac.SemanticdbPlugin
+import scala.meta.io.AbsolutePath
+import scala.meta.io.Classpath
+import scalafix.interfaces.ScalafixDiagnostic
+import scalafix.interfaces.ScalafixException
+import scalafix.interfaces.ScalafixMainCallback
+import scalafix.interfaces.ScalafixMainMode
+import scalafix.internal.reflect.ClasspathOps
+import scalafix.internal.reflect.RuleCompiler
+import scalafix.internal.sbt.ScalafixJarFetcher
+import scalafix.testkit.DiffAssertions
+
+class ScalafixImplSuite extends FunSuite with DiffAssertions {
+  def semanticdbPluginPath(): String = {
+    // Copy-pasted code from metac command-line tool
+    // https://github.com/scalameta/scalameta/blob/9f15793aae3cb6a00e1e2d6bcbf13e9c234ea91f/semanticdb/metac/src/main/scala/scala/meta/internal/metac/Main.scala#L14-L23
+    val manifestDir = Files.createTempDirectory("semanticdb-scalac_")
+    val resourceUrl =
+      classOf[SemanticdbPlugin].getResource("/scalac-plugin.xml")
+    val resourceChannel = Channels.newChannel(resourceUrl.openStream())
+    val manifestStream = new FileOutputStream(
+      manifestDir.resolve("scalac-plugin.xml").toFile)
+    manifestStream.getChannel.transferFrom(resourceChannel, 0, Long.MaxValue)
+    manifestStream.close()
+    val pluginClasspath = classOf[SemanticdbPlugin].getClassLoader match {
+      case null => manifestDir.toString
+      case cl: URLClassLoader =>
+        cl.getURLs.map(_.getFile).mkString(File.pathSeparator)
+      case cl => sys.error(s"unsupported classloader: $cl")
+    }
+    pluginClasspath
+  }
+  def scalaLibrary: AbsolutePath =
+    RuleCompiler.defaultClasspathPaths
+      .find(_.toNIO.getFileName.toString.contains("scala-library"))
+      .getOrElse {
+        throw new IllegalStateException("Unable to detect scala-library.jar")
+      }
+
+  test("versions") {
+    val api = i.Scalafix.classloadInstance(this.getClass.getClassLoader)
+    assert(api.scalafixVersion() == Versions.version)
+    assert(api.scalametaVersion() == Versions.scalameta)
+    assert(api.scala211() == Versions.scala211)
+    assert(api.scala212() == Versions.scala212)
+    assert(
+      api
+        .supportedScalaVersions()
+        .sameElements(Versions.supportedScalaVersions)
+    )
+    val help = api.mainHelp(80)
+    assert(help.contains("Usage: scalafix"))
+  }
+
+  test("error") {
+    val cl = new URLClassLoader(Array())
+    val ex = intercept[ScalafixException] {
+      i.Scalafix.classloadInstance(cl)
+    }
+    assert(ex.getCause.isInstanceOf[ClassNotFoundException])
+  }
+
+  test("runMain") {
+    // This is a full integration test that stresses the full breadth of the scalafix-interfaces API
```

Comment:
Here is a full integration test for the new reflective API. Gives an idea how it will be used.